File: apps/client/local_plugins/translate_server/ios/Classes/AgentSinkAdapter.swift
import Foundation
import ai_plugin_interface

/// 把翻译型 agent 的 [AgentEventSink] 事件转译成通话翻译统一字幕/状态/错误事件。
final class AgentSinkAdapter: AgentEventSink {

    private let sessionId: String
    private let leg: CallLeg
    private let emit: ([String: Any?]) -> Void

    private let readyLock = NSLock()
    private let readySem = DispatchSemaphore(value: 0)
    private var readyError: TranslateSessionError?
    private var readyArrived = false

    init(sessionId: String, leg: CallLeg, emit: @escaping ([String: Any?]) -> Void) {
        self.sessionId = sessionId
        self.leg = leg
        self.emit = emit
    }

    func awaitReady(timeout: TimeInterval) throws {
        readyLock.lock()
        if readyArrived {
            let err = readyError
            readyLock.unlock()
            if let err = err { throw err }
            return
        }
        readyLock.unlock()

        let result = readySem.wait(timeout: .now() + timeout)
        readyLock.lock()
        defer { readyLock.unlock() }
        if result == .timedOut && !readyArrived {
            readyArrived = true
            readyError = .connectTimeout
            throw TranslateSessionError.connectTimeout
        }
        if let err = readyError { throw err }
    }

    func onSttEvent(_ event: SttEventData) {
        switch event.kind {
        case "partialResult":
            emit(TranslateEvents.subtitle(
                sessionId: sessionId, leg: leg.rawValue, stage: "partial",
                sourceText: event.text ?? "", translatedText: nil, requestId: nil
            ))
        case "finalResult":
            guard !event.requestId.isEmpty else { return }
            emit(TranslateEvents.subtitle(
                sessionId: sessionId, leg: leg.rawValue, stage: "final",
                sourceText: event.text ?? "", translatedText: nil, requestId: event.requestId
            ))
        case "error":
            emit(TranslateEvents.error(
                sessionId: sessionId,
                code: event.errorCode ?? "stt.error",
                message: event.errorMessage ?? "stt error",
                leg: leg.rawValue
            ))
        default: break
        }
    }

    func onLlmEvent(_ event: LlmEventData) {
        switch event.kind {
        case "firstToken":
            guard !event.requestId.isEmpty else { return }
            emit(TranslateEvents.subtitle(
                sessionId: sessionId, leg: leg.rawValue, stage: "partial",
                sourceText: "", translatedText: event.textDelta ?? "", requestId: event.requestId
            ))
        case "done":
            guard !event.requestId.isEmpty else { return }
            emit(TranslateEvents.subtitle(
                sessionId: sessionId, leg: leg.rawValue, stage: "final",
                sourceText: "", translatedText: event.fullText ?? event.textDelta ?? "",
                requestId: event.requestId
            ))
        case "error":
            emit(TranslateEvents.error(
                sessionId: sessionId,
                code: event.errorCode ?? "llm.error",
                message: event.errorMessage ?? "llm error",
                leg: leg.rawValue
            ))
        default: break
        }
    }

    func onTtsEvent(_ event: TtsEventData) {
        if event.kind == "error" {
            emit(TranslateEvents.error(
                sessionId: sessionId,
                code: event.errorCode ?? "tts.error",
                message: event.errorMessage ?? "tts error",
                leg: leg.rawValue
            ))
        }
    }

    func onStateChanged(sessionId: String, state: String, requestId: String?) {}

    func onError(sessionId: String, errorCode: String, message: String, requestId: String?) {
        emit(TranslateEvents.error(
            sessionId: self.sessionId, code: errorCode, message: message,
            leg: leg.rawValue
        ))
    }

    func onConnectionStateChanged(sessionId: String, state: String, errorMessage: String?) {
        emit(TranslateEvents.connectionState(
            sessionId: self.sessionId, leg: leg.rawValue,
            state: state, errorMessage: errorMessage
        ))
    }

    func onAgentReady(sessionId: String, ready: Bool, errorCode: String?, errorMessage: String?) {
        readyLock.lock()
        if !readyArrived {
            readyArrived = true
            if !ready {
                readyError = .agentReadyFailed(
                    code: errorCode ?? "agent.ready_failed",
                    message: errorMessage ?? "agent ready failed",
                    leg: leg.rawValue
                )
            }
            readyLock.unlock()
            readySem.signal()
        } else {
            readyLock.unlock()
        }

        if !ready {
            emit(TranslateEvents.error(
                sessionId: self.sessionId,
                code: errorCode ?? "agent.ready_failed",
                message: errorMessage ?? "agent ready failed",
                leg: leg.rawValue, fatal: true
            ))
        }
    }
}

enum TranslateSessionError: Error, CustomStringConvertible {
    case noDevice(message: String)
    case connectTimeout
    case agentUnsupported(message: String)
    case enterFailed(message: String)
    case agentReadyFailed(code: String, message: String, leg: String)

    var code: String {
        switch self {
        case .noDevice: return "translate.no_device"
        case .connectTimeout: return "translate.connect_timeout"
        case .agentUnsupported: return "translate.agent_unsupported"
        case .enterFailed: return "translate.enter_mode_failed"
        case .agentReadyFailed(let c, _, _): return c
        }
    }

    var description: String {
        switch self {
        case .noDevice(let m): return "translate.no_device: \(m)"
        case .connectTimeout: return "translate.connect_timeout: agent connect timeout"
        case .agentUnsupported(let m): return "translate.agent_unsupported: \(m)"
        case .enterFailed(let m): return "translate.enter_mode_failed: \(m)"
        case .agentReadyFailed(let c, let m, let leg): return "\(c) [\(leg)]: \(m)"
        }
    }
}

File: apps/client/local_plugins/translate_server/android/src/main/kotlin/com/aiagent/translate_server/AgentSinkAdapter.kt
package com.aiagent.translate_server

import com.aiagent.plugin_interface.AgentEventSink
import com.aiagent.plugin_interface.LlmEventData
import com.aiagent.plugin_interface.SttEventData
import com.aiagent.plugin_interface.TtsEventData
import kotlinx.coroutines.CompletableDeferred

/**
 * 把翻译型 agent 的 [AgentEventSink] 事件转译成通话翻译统一的字幕/状态/错误事件。
 *
 * 配对策略（按 [requestId]）：
 *  - 每条 STT finalResult / LLM done 事件携带 `requestId`，UI 端按 `requestId`
 *    把 source 和 translated 配对到同一行字幕，**不依赖到达顺序**。
 *  - SttEventData(partialResult) → stage=partial, sourceText, requestId=null（未定稿）
 *  - SttEventData(finalResult)   → stage=final,   sourceText, translatedText=null
 *  - LlmEventData(firstToken)    → stage=partial, translatedText, requestId（更新对应行的译文 partial）
 *  - LlmEventData(done)          → stage=final,   translatedText（更新对应行的译文 final）
 *
 * 这样：连续多句翻译时，UI 不会把旧句的译文错位到新句的源文下。
 */
internal class AgentSinkAdapter(
    private val sessionId: String,
    private val leg: CallLeg,
    private val emit: (Map<String, Any?>) -> Unit,
    /** 第一次 onAgentReady 时 complete（成功）或 completeExceptionally（失败），编排器据此等待就绪。 */
    val connected: CompletableDeferred<Unit> = CompletableDeferred(),
) : AgentEventSink {

    override fun onSttEvent(event: SttEventData) {
        when (event.kind) {
            "partialResult" -> {
                // 进行中的源文（无 requestId），UI 显示 in-progress 当前句的源文。
                emit(TranslateEvents.subtitle(
                    sessionId = sessionId,
                    leg = leg.wireName,
                    stage = "partial",
                    sourceText = event.text ?: "",
                    translatedText = null,
                    requestId = null,
                ))
            }
            "finalResult" -> {
                // 一句源文定稿，按 requestId 钉到 UI 端。译文留 null，由后续 LLM done 补。
                val reqId = event.requestId.takeIf { it.isNotBlank() } ?: return
                emit(TranslateEvents.subtitle(
                    sessionId = sessionId,
                    leg = leg.wireName,
                    stage = "final",
                    sourceText = event.text ?: "",
                    translatedText = null,
                    requestId = reqId,
                ))
            }
            "error" -> emit(TranslateEvents.error(
                sessionId = sessionId,
                code = event.errorCode ?: "stt.error",
                message = event.errorMessage ?: "stt error",
                leg = leg.wireName,
            ))
            // listeningStarted / listeningStopped / vadSpeechStart/end 此处不用透传
        }
    }

    override fun onLlmEvent(event: LlmEventData) {
        when (event.kind) {
            "firstToken" -> {
                val reqId = event.requestId.takeIf { it.isNotBlank() } ?: return
                emit(TranslateEvents.subtitle(
                    sessionId = sessionId,
                    leg = leg.wireName,
                    stage = "partial",
                    sourceText = "",            // 源文已在 STT final 时定稿；这里不复发
                    translatedText = event.textDelta ?: "",
                    requestId = reqId,
                ))
            }
            "done" -> {
                val reqId = event.requestId.takeIf { it.isNotBlank() } ?: return
                emit(TranslateEvents.subtitle(
                    sessionId = sessionId,
                    leg = leg.wireName,
                    stage = "final",
                    sourceText = "",
                    translatedText = event.fullText ?: event.textDelta ?: "",
                    requestId = reqId,
                ))
            }
            "error" -> emit(TranslateEvents.error(
                sessionId = sessionId,
                code = event.errorCode ?: "llm.error",
                message = event.errorMessage ?: "llm error",
                leg = leg.wireName,
            ))
        }
    }

    override fun onTtsEvent(event: TtsEventData) {
        // TTS 字节通过 ExternalAudioSink 拿，不走这里。事件里只关心 error。
        if (event.kind == "error") {
            emit(TranslateEvents.error(
                sessionId = sessionId,
                code = event.errorCode ?: "tts.error",
                message = event.errorMessage ?: "tts error",
                leg = leg.wireName,
            ))
        }
    }

    override fun onStateChanged(sessionId: String, state: String, requestId: String?) {
        // agent 内部状态机：idle/connected/listening/processing... 通话翻译不消费。
    }

    override fun onError(sessionId: String, errorCode: String, message: String, requestId: String?) {
        emit(TranslateEvents.error(
            sessionId = this.sessionId,
            code = errorCode,
            message = message,
            leg = leg.wireName,
        ))
    }

    override fun onConnectionStateChanged(sessionId: String, state: String, errorMessage: String?) {
        emit(TranslateEvents.connectionState(
            sessionId = this.sessionId,
            leg = leg.wireName,
            state = state,
            errorMessage = errorMessage,
        ))
        // 注意：就绪/失败信号统一走 onAgentReady；这里只负责把状态向上 emit 给 Flutter。
    }

    override fun onAgentReady(sessionId: String, ready: Boolean, errorCode: String?, errorMessage: String?) {
        if (ready) {
            if (!connected.isCompleted) connected.complete(Unit)
        } else {
            if (!connected.isCompleted) {
                connected.completeExceptionally(
                    IllegalStateException(
                        "agent ${leg.wireName} ready failed: ${errorCode ?: "unknown"} ${errorMessage ?: ""}"
                    )
                )
            }
            emit(TranslateEvents.error(
                sessionId = this.sessionId,
                code = errorCode ?: "agent.ready_failed",
                message = errorMessage ?: "agent ready failed",
                leg = leg.wireName,
                fatal = true,
            ))
        }
    }

}
